WIP: Current work on Variant specs (#461)
apache/parquet-format: VariantEncoding.md

For example, in a Variant containing an Array of Variant values, the representation of an inner Variant value, when paired with the metadata of the full variant, is itself a valid Variant.

This document describes the Variant Binary Encoding scheme.
The [Variant Shredding spec](VariantShredding.md) describes the details of shredding Variant values as typed Parquet columns.

## Variant in Parquet

> A Variant value in Parquet is represented by a group with 2 fields, named `variant_value` and `metadata`.

Reviewer: @rdblue I think changing to `variant_value` could be trickier, since Spark has already written out data with `value` + `metadata`.

Instead, I proposed changing it the other direction, renaming the shredding one to `value`. That change is part of the change to add the logical type here: https://github.com/apache/parquet-format/pull/460.

What do you think?

Author: Yeah, that sounds correct to me. I'll update it here.

The Variant group must be annotated with the `VARIANT` logical type.
Both fields `value` and `metadata` are of type `binary`.
The `metadata` field is required and must be a valid Variant metadata, as defined below.
The `value` field is optional.
When present, the `value` field must be a valid Variant value, as defined below. 
The `value` field may be null only when parts of the Variant value are shredded according to the Variant Shreedding spec.

This is the expected representation in Parquet:

```
optional group variant_event (VARIANT) {
  required binary metadata;

>   optional binary value;

Reviewer 2: Would an engine be violating the spec if it is not shredding and hence declaring the value field as required?

Author: No. The encoding spec says that the value must be required when not shredding. I didn't include that case here since this is documenting how to shred variants.

Author: Sorry, I forgot I was updating the encoding spec here. I'm updating this to match the `LogicalTypes.md` PR, which uses `required` for non-shredded Variant.

For example, in a Variant containing an Array of Variant values, the representation of an inner Variant value, when paired with the metadata of the full variant, is itself a valid Variant.

This document describes the Variant Binary Encoding scheme.

> The [Variant Shredding specification](VariantShredding.md) describes the details of shredding Variant values as typed Parquet columns.

Reviewer 3: Varia
```suggestion
Variant fields can also be "shredded".  Shredding refers to extracting some elements of the variant into separate columns for more efficient extraction/filter pushdown. The [Variant Shredding specification](VariantShredding.md) describes the details of shredding Variant values as typed Parquet columns.
```

I think this is the first place shredding is used and I think we should try to give readers a background on what is meant here?

Author: Thanks! Updated.

## Variant in Parquet

A Variant value in Parquet is represented by a group with 2 fields, named `value` and `metadata`.

> * The Variant group must be annotated with the `VARIANT` logical type.

Author: Size limit of what?

> * The Variant group must be annotated with the `VARIANT` logical type.

Reviewer 4: When Spark implements the Variant, internally it limits value and metadata to 16MB so it will have the limit on how much data can be stored. Different engines may have different limits internally and some Variant values may not be loaded across the engines. I'm wondering if we should have such limit in the spec explicitly. But of course it's really hard to really define such kind of limit since the input data of same size may be encoded differently.

> * The Variant group must be annotated with the `VARIANT` logical type.

Reviewer 3: I think engines are always going to have different limits on field sizes, I think providing a recommendation on size makes sense but I don't think it should be a rquirement.

> * The Variant group must be annotated with the `VARIANT` logical type.

Author: I agree with Micah and Gene. Let's not put a limit in the spec. That should be determined in other ways.

* Both fields `value` and `metadata` must be of type `binary` (called `BYTE_ARRAY` in the Parquet thrift definition).
* The `metadata` field is required and must be a valid Variant metadata, as defined below.

> * The `value` field is required for unshredded Variant values.

Reviewer 3: ```suggestion
* The `value` must be annotated either as required (no shredding) or optional (if shredding might happen)
```
This might address the confusion on @wgtmac comment below (and we eliminate the other bullet points below, if we inline the link.

> * The `value` field is optional when parts of the Variant value are shredded according to the [Variant Shredding specification](VariantShredding.md).

Author: @wgtmac, I don't think this makes sense. A Variant value may have any structure. We cannot know ahead of time (when setting the Parquet type) that all parts of all encoded Variant values will be shredded.

> * The `value` field is optional when parts of the Variant value are shredded according to the [Variant Shredding specification](VariantShredding.md).

Reviewer 3: I think the  confusion might be around what this is intended to mean. Whether the column can be left out entirely, or whether we are talking about the cardinality specifier of (required/optional/repeated).  Maybe we can clarify this as saying.

If so we can maybe clarify that the field maybe be annotated as required or optional.  When it is annoted as required there is no shredding. It only optional to when shredding might be used?

> * The `value` field is optional when parts of the Variant value are shredded according to the [Variant Shredding specification](VariantShredding.md).

Reviewer 3: (or maybe clarify that field must always be present but the cardinality changes depending on if the field is shredded).

* When present, the `value` field must be a valid Variant value, as defined below. 

This is the expected unshredded representation in Parquet:

```
optional group variant_name (VARIANT) {
  required binary metadata;
  required binary value;
}
```

> This is an example representation of a shredded Variant in Parquet:

Reviewer 3: it might pay to describe what this is (typed_value stores the value of the variant when it is a simple integer?) and given another pointer to variant

Author: This already points to the shredding spec in multiple places, so I think it is clear how to get more information about `typed_value`.

Reviewer 3: Added concrete suggestion, this is more about flow, and what the reader is expected to understand at this point.

```
optional group shredded_variant_name (VARIANT) {
  required binary metadata;
  optional binary value;
  optional int64 typed_value;
}
```

> There are no restrictions on the repetition of Variant groups (required, optional, or repeated).

Reviewer 3: shouldn't repeated use 3 level list structure?

Author: This is saying that the Variant type doesn't place _additional_ restrictions. I'll update to this:
> The `VARIANT` annotation places no additional restrictions on the repetition of Variant groups, but repetition may be restricted by containing types (such as `MAP` and `LIST`).

> The `VARIANT` annotation places no additional restrictions on the repetition of Variant groups, but repetition may be restricted by containing types (such as `MAP` and `LIST`).

Reviewer 3: ```suggestion
The `VARIANT` type, despite being a group, is considered a primitive type, and therefore places no additional restrictions on the repetition of Variant groups, but repetition may be restricted by containing types (such as `MAP` and `LIST`).
```

Author: I don't agree that it is considered a primitive type. And we don't need to in order to state that it places no additional restrictions on the repetition of Variant groups.

Reviewer 3: OK, I don't think this matters too much.

This document describes the Variant Binary Encoding scheme.
Variant fields can also be _shredded_.
Shredding refers to extracting some elements of the variant into separate columns for more efficient extraction/filter pushdown.
The [Variant Shredding specification](VariantShredding.md) describes the details of shredding Variant values as typed Parquet columns.

## Variant in Parquet

A Variant value in Parquet is represented by a group with 2 fields, named `value` and `metadata`.

* The Variant group must be annotated with the `VARIANT` logical type.
* Both fields `value` and `metadata` must be of type `binary` (called `BYTE_ARRAY` in the Parquet thrift definition).
* The `metadata` field is `required` and must be a valid Variant metadata, as defined below.
* The `value` field must be annotated as `required` for unshredded Variant values, or `optional` if parts of the value are [shredded](VariantShredding.md) as typed Parquet columns.
* When present, the `value` field must be a valid Variant value, as defined below. 

This is the expected unshredded representation in Parquet:

```
optional group variant_name (VARIANT) {

>   required binary metadata;

Reviewer 5: @chenhao-db @cashmand Hi, I noticed that currently Spark's Variant writes the value first and then the metadata.  which is the opposite of shredding. Have we considered making adjustments to this?

spark's variant
```
optional group variant_name (VARIANT) {
  required binary value;
  required binary metadata;
}
```

Reviewer 6: Hi @Zouxxyy, in the Spark shredding PRs I've been working on, I put metadata first. I didn't see much benefit to changing the order in the existing non-shredded code, but I don't feel too strongly about it either way. The spec is pretty clear that readers should identify the appropriate columns based on field names, not field order, and I think things could become quite fragile if they did rely on field order.

Reviewer 5: > Hi @Zouxxyy, in the Spark shredding PRs I've been working on, I put metadata first. I didn't see much benefit to changing the order in the existing non-shredded code, but I don't feel too strongly about it either way. The spec is pretty clear that readers should identify the appropriate columns based on field names, not field order, and I think things could become quite fragile if they did rely on field order.

Thank you, but future users or developers may find it strange if the actual implementation differs from the specs. I think it's better to adhere to the specs as long as it doesn’t affect performance, before the official release of spark 4.0. If you don’t mind, I can work on this and raise a PR to spark, WDYT

Reviewer 6: Okay, I think it's fine if you want to make a PR with the change.

Reviewer 5: > Okay, I think it's fine if you want to make a PR with the change.

Oh, I checked again, and it seems that `new VariantVal(byte[] value, byte[] metadata)` is already used everywhere in spark, this change comes at a significant cost. @cashmand @rdblue , do you think the inconsistency here with the specs will have a big impact in the future?

The first `offset` value will always be `0`, and the last `offset` value will always be the total length of `bytes`.
The last part of the metadata is `bytes`, which stores all the string values in the dictionary.

> ### Metadata encoding grammar

Author: I've updated these to not use more than one H1, which can cause issues with TOC. Pages should have just one H1.

Reviewer 7: TIL as well

> | Logical Type         | Physical Type               | Type ID | Equivalent Parquet Type     | Binary format                                                                                                       |

Reviewer 4: @gene-db  When we say Logical Type and Physical Type here, what are we exactly referring to? Should we refer to Parquet logical type and Parquet physical type?

Author: Updated to clarify the columns that are Variant types.

For the Parquet types, I'd prefer to keep it as is to avoid a larger table. Even though it mixes logical and physical types, it is unambiguous.

## Shredding

A single Variant object may have poor read performance when only a small subset of fields are needed.
A better approach is to create separate columns for individual fields, referred to as shredding or subcolumnarization.
[VariantShredding.md](VariantShredding.md) describes the Variant shredding specification in Parquet.

## Conversion to JSON

Values stored in the Variant encoding are a superset of JSON values.
For example, a Variant value can be a date that has no equivalent type in JSON.
To maximize compatibility with readers that can process JSON but not Variant, the following conversions should be used when producing JSON from a Variant:

| Variant type  | JSON type | Representation requirements                              | Example                              |
|---------------|-----------|----------------------------------------------------------|--------------------------------------|
| Null type     | null      | `null`                                                   | `null`                               |
| Boolean       | boolean   | `true` or `false`                                        | `true`                               |

> | Exact Numeric | number    | Digits in fraction must match scale, no exponent         | `34`, 34.00                          |

Reviewer: For exact numerics, we should allow truncating trailing zeros. For example, `int8` value `1` and `decimal(5,2)` value `100` can both be represented as a JSON value `1`.

Also, should the example be quoted to stay consistent?
```suggestion
| Exact Numeric | number    | Digits in fraction must match scale, no exponent         | `34`, `34.00`                          |
```

Reviewer 6: I think the intent of considering `Exact Numeric` to be a single logical type is that we consider the `int8` value `1` to be logically equivalent to `decimal(5,2)` with unscaled value `100`. If that's the case, I think we'd want the produced JSON to be the same for both (probably `1` in both cases), and not recommend having the fraction match the scale.

Author: @gene-db, @cashmand, these are concerns for the engine layer, not for storage. If Spark wants to automatically coerce between types that's fine, but the compromise that we talked about a couple months ago was to leave this out of the shredding spec and delegate the behavior to engines. Storage should always produce the data that was stored, without modification.

Reviewer: Yes, the engine should be the one concerned with changing types.

However, my original question was about this JSON representation wording. Currently, the `Representation requirements` for an `Exact Numeric` says the `Digits in fraction must match scale`. However, because the `Exact Numeric` is considered a logical type, the value `1` could be stored in the Variant as `int8` 1 or `decimal(5,2)` 100. Both of those would be the same numeric value, so we should allow truncating trailing zeros in the JSON representation, instead of requiring that the digits in the fraction match the scale.

Reviewer: @rdblue I am confused with this JSON chart then. If we are talking about "storage implementation", then are you expecting there is a "storage implementation" that is converting variant values to JSON? When will storage convert a variant value to a JSON string?

I originally thought this chart was trying to say, "When an engine wants to convert a variant value to a JSON string, here are the rules". Therefore, we should allow engines to cast integral decimals to integers before converting to JSON, as you already mentioned in your previous comment.

Reviewer 3: I think I agree with @gene-db on this.  I think any json representation that has semantically the same meaning in JSON should be allowed.  Translation to JSON is inherently lossy and I think trying to match semantics will be more error prone then it is worth (i.e. it  should be a non-goal to expect it to be able to reconstruct the exact same variant from the proposed JSON representation).

Author: > When an engine wants to convert a variant value to a JSON string, here are the rules

Yes, this is correct. We want a clear way to convert to a JSON string. However, the normalization needs to happen first. We don't want to specify that the JSON must be any more lossy than it already is.

Why would we require an engine to produce a normalized value?

Reviewer 3: > Why would we require an engine to produce a normalized value?

At least for me, I don't think it is about "requiring" and engine to produce a normalized value first.  I think if an engine is reading variant and converting it to JSON, it is possibly doing so through an internal representation so it can still apply operators on top of the JSON value and possibly even storing it as an internal representation.  Conversion to a string is really only an end-user visible thing.  So when I read this it seems to be requiring an engine to NOT normalize which could be hard to implement for some engines.

> | Float         | number    | Fraction must be present                                 | `14.20`                              |

Reviewer 3: we should cover expected format for +/- inf and NaN

Reviewer 3: Seems reasonable to me.

| Double        | number    | Fraction must be present                                 | `1.0`                                |
| Date          | string    | ISO-8601 formatted date                                  | `"2017-11-16"`                       |
| Timestamp     | string    | ISO-8601 formatted UTC timestamp including +00:00 offset | `"2017-11-16T22:31:08.000001+00:00"` |

> | TimestampNTZ  | string    | ISO-8601 formatted UTC timestamp with no offset or zone  | `"2017-11-16T22:31:08.000001"`       |

Reviewer 3: What precision decimal values are required?

Author: We can either use 6 or infer 0s to fill in. I don't think it matters much. What do you prefer?

Reviewer 3: I don't have a strong preference.  I think truncation is probably fine (I think this mainly becomes question for when nanos are added if we want to disambiguate them).

A better approach is to create separate columns for individual fields, referred to as shredding or subcolumnarization.
[VariantShredding.md](VariantShredding.md) describes the Variant shredding specification in Parquet.

## Conversion to JSON

Values stored in the Variant encoding are a superset of JSON values.
For example, a Variant value can be a date that has no equivalent type in JSON.
To maximize compatibility with readers that can process JSON but not Variant, the following conversions should be used when producing JSON from a Variant:

| Variant type  | JSON type | Representation requirements                              | Example                              |
|---------------|-----------|----------------------------------------------------------|--------------------------------------|
| Null type     | null      | `null`                                                   | `null`                               |
| Boolean       | boolean   | `true` or `false`                                        | `true`                               |
| Exact Numeric | number    | Digits in fraction must match scale, no exponent         | `34`, 34.00                          |

> | Float         | number    | Fraction must be present                                 | `14.20`                              |

Reviewer 3: it might be worth noting that relying on the json values here might be lossy for some numbers (this is implied but people still ask questions ...)

| Double        | number    | Fraction must be present                                 | `1.0`                                |
| Date          | string    | ISO-8601 formatted date                                  | `"2017-11-16"`                       |
| Timestamp     | string    | ISO-8601 formatted UTC timestamp including +00:00 offset | `"2017-11-16T22:31:08.000001+00:00"` |
| TimestampNTZ  | string    | ISO-8601 formatted UTC timestamp with no offset or zone  | `"2017-11-16T22:31:08.000001"`       |

> | Binary        | string    | Base64 encoded binary                                    | `"dmFyaWFudAo="`                     |

Reviewer 3: I guess the other alternative here is a hex encoded string which will always be bigger, base64 seems reasonable just want to call this out in case people have preferences.
